src/python/pants/backend/python/tasks/build_local_python_distributions.py

```diff
@@ -44,89 +45,103 @@ def product_types(cls):
   @classmethod
   def prepare(cls, options, round_manager):
     round_manager.require_data(PythonInterpreter)
+    round_manager.require(SharedLibrary)
 
   @classmethod
   def implementation_version(cls):
-    return super(BuildLocalPythonDistributions, cls).implementation_version() + [('BuildLocalPythonDistributions', 2)]
+    return super(BuildLocalPythonDistributions, cls).implementation_version() + [('BuildLocalPythonDistributions', 3)]
 
   @classmethod
   def subsystem_dependencies(cls):
-    return super(BuildLocalPythonDistributions, cls).subsystem_dependencies() + (NativeToolchain.scoped(cls),)
+    return super(BuildLocalPythonDistributions, cls).subsystem_dependencies() + (
+      PythonNativeCode.scoped(cls),
+    )
 
-  @memoized_method
-  def _native_toolchain_instance(self):
-    return NativeToolchain.scoped_instance(self)
+  @memoized_property
+  def _python_native_code_settings(self):
+    return PythonNativeCode.scoped_instance(self)
 
+  # FIXME(#5869): delete this and get Subsystems from options, when that is possible.
+  def _request_single(self, product, subject):
+    # NB: This is not supposed to be exposed to Tasks yet -- see #4769 to track the status of
+    # exposing v2 products in v1 tasks.
+    return self.context._scheduler.product_request(product, [subject])[0]
+
+  @memoized_property
+  def _setup_py_native_tools(self):
+    native_toolchain = self._python_native_code_settings.native_toolchain
+    return self._request_single(SetupPyNativeTools, native_toolchain)
+
+  # TODO: This should probably be made into an @classproperty (see PR #5901).
   @property
   def cache_target_dirs(self):
     return True
 
-  @staticmethod
-  def filter_target(tgt):
-    return type(tgt) is PythonDistribution
+  def _get_setup_requires_to_resolve(self, dist_target):
+    if not dist_target.setup_requires:
+      return None
 
-  def _ensure_setup_requires_site_dir(self, dist_targets, interpreter, site_dir):
     reqs_to_resolve = set()
 
-    for tgt in dist_targets:
-      for setup_req_lib_addr in tgt.setup_requires:
-        for req_lib in self.context.build_graph.resolve(setup_req_lib_addr):
-          for req in req_lib.requirements:
-            reqs_to_resolve.add(req)
+    for setup_req_lib_addr in dist_target.setup_requires:
+      for req_lib in self.context.build_graph.resolve(setup_req_lib_addr):
+        for req in req_lib.requirements:
+          reqs_to_resolve.add(req)
 
-    if not reqs_to_resolve:
+    if len(reqs_to_resolve) == 0:
```

Comment:
Why switch this?

Reply:
No reason, reverted. Thanks, and please point out more issues like this, it's not intentional and I should have found this when going over the diff.